File: Microsoft/microsoft/2/2.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def reverse(self):
        prev = None
        current = self.head
        while (current is not None):
            next = current.next
            current.next = prev
            prev = current
            current = next
        self.head = prev

    def push(self, new_data):
        new_node = Node(new_data)
        new_node.next = self.head
        self.head = new_node

    def printList(self):
        temp = self.head
        while (temp):
            print
            temp.data,
            temp = temp.next



llist = LinkedList()
n = int(input());
arr = []
arr = [x for x in input().split()]
for i in arr:
    llist.push(int(i))

llist.printList()
llist.reverse()
llist.printList()

File: Microsoft/2/2.cpp
#include <iostream> 
using namespace std; 

struct Node { 
    int data; 
    struct Node* next; 
    Node(int data) 
    { 
        this->data = data; 
        next = NULL; 
    } 
}; 
  struct LinkedList { 
    Node* head; 
    LinkedList() 
    { 
        head = NULL; 
    } 
    void reverse() 
    { 
        Node* current = head; 
        Node *prev = NULL, *next = NULL; 
  
        while (current != NULL) { 
            next = current->next; 
            current->next = prev; 
            prev = current; 
            current = next; 
        } 
        head = prev; 
    } 
    void print() 
    { 
        struct Node* temp = head; 
        while (temp != NULL) { 
            cout << temp->data << " "; 
            temp = temp->next; 
        } 
    } 
  
    void push(int data) 
    { 
        Node* temp = new Node(data); 
        temp->next = head; 
        head = temp; 
    } 
}; 
int main() 
{ 
    LinkedList ll; 
    int n;
    cin>>n;
    for(int i=0;i<n;i++){
        int x;
        cin>>x;
    ll.push(x);
    } 
    ll.print(); cout<<endl;
    ll.reverse(); 
    ll.print(); 
    return 0; 
} 
